rpc/src/router.rs

```diff
@@ -0,0 +1,153 @@
+use std::{collections::HashMap, marker::PhantomData};
+
+use serde::{Deserialize, Serialize};
+
+pub struct Router {
+    pub routes: HashMap<String, Box<dyn Handler + 'static + Send + Sync>>,
+}
+
+impl Router {
+    pub fn new() -> Self {
+        Router {
+            routes: HashMap::new(),
+        }
+    }
+
+    pub fn from_routes(
+        routes: HashMap<String, Box<dyn Handler + 'static + Send + Sync>>,
+    ) -> Self {
+        Router { routes }
+    }
+
+    pub fn add<Marker: 'static + Send + Sync>(
+        &mut self,
+        name: &str,
+        function: impl HandlerFunction<Marker>,
+    ) {
+        self.routes.insert(
+            name.to_owned(),
+            Box::new(FunctionHandler {
+                function,
+                marker: PhantomData,
+            }),
+        );
+    }
+
+    pub fn call(&self, name: &str, args: Vec<String>) -> String {
+        match self.routes.get(name) {
+            Some(handler) => handler.call(args),
+            None => NOT_FOUND.into(),
+        }
+    }
+}
+
+impl Default for Router {
+    fn default() -> Self {
+        Router::new()
+    }
+}
+
+#[derive(Serialize)]
+pub struct Response<T> {
+    pub result: Option<T>,
+    pub error: Option<String>,
+    pub is_success: bool,
+}
+
+const CATASTROPHIC_ERROR: &str = "{\"result\": null, \"error\": \"CATASTROPHIC_ERROR: Failed to serialize response\", \"is_success\": false}";
+const NOT_FOUND: &str = "{\"result\": null, \"error\": \"NOT_FOUND: Unknown function\", \"is_success\": false}";
+
+impl<T> Response<T> {
+    pub fn success(result: T) -> Self {
+        Response {
+            result: Some(result),
+            error: None,
+            is_success: true,
+        }
+    }
+
+    pub fn error(error: String) -> Self {
+        Response {
+            result: None,
+            error: Some(error),
+            is_success: false,
+        }
+    }
+}
+pub trait HandlerFunction<Marker>: Send + Sync + 'static {
```

Comment:
Renaming this to `Callback` will be more appropriate because it represents a function that will be boxed and put on the heap and called with some value.

Moreover, having a different name will reduce confusion between it and FunctionHandler.